### DISCUSSION TITLE
Subscribe to native events in other native modules

### DISCUSSION THREAD

**Discussion by a1lu** [react-native-community/discussions-and-proposals#363] [Ideas]
# Feature Request

## Why it is needed

Modules that send a lot of events (in my case bluetooth notifications) put a lot load on the js bridge. If you don't need this data in javascript code but only in another native module this load is unnecessary. 
If you want to use the existing modules environment and don't want to reinvent the wheel there is AFAIK no way to avoid this.
A better way would be to subscribe the events from the native module and keep the data in native land.

## Possible implementation

The subscriber could subscribe the event in the react context object. If another modules sends a event, this event could be queued in a blocking queue for each subscriber. The subscriber can than fetch the event in a thread or poll this queue.

E.g. 
```
// subscriber
BlockingQueue queue;
public class MyThread implements Runnable
{
  @Override public void run()
  {
    while(true) {
     // fetch data from blocking queue
     // process data
    }
  }
}


reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
      .subscribe(eventName, queue);


// sender
 reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
      .emit(eventName, params);

```
I am not an android or iOS expert, so maybe there is a better solution to send events in native. 

In my project I patched a 3rd party bluetooth library I am using to put the event parameters into a global queue (android only) and my modules fetches the data in a thread like shown above. This reduced the cpu load by 5 to 10% on a Moto G5.

Best regards

**Comment by kelset** [comment]
This sounds interesting - btw, have you looked into the TurboModules #40?

**Comment by a1lu** [comment]
Not yet.
On the first glance that looks like better native modules, though how is this related to the communication between modules?


Edit:
In my case I removed the emit() call and replaced it by the queue. I think if emit() is used, the event is always propagated to JS? So maybe we should check if someone has subscribed the event before sending it though the bridge? I program gets 20Bytes every 5ms, this used almost 90% of the bridge on a Moto G5, on older devices it did not work. I am not sure about all consequences this check would have (e.g. the bridge spy tool would not see any activity).

**Comment by terrysahaidak** [comment]
@a1lu TurboModules won't block the js thread, because there is "no bridge" (it is but in some different way).

Also, I didn't get why did you use `RCTDeviceEventEmitter` for sending events between native modules. It's a tool to send events via the bridge.

It would better to try to use some more native approach for both android and ios -  some kind of event emitter. I'm not an expert, but I'm pretty sure it exists for both platforms.

**Comment by a1lu** [comment]
@terrysahaidak 
> Also, I didn't get why did you use `RCTDeviceEventEmitter` for sending events between native 


I'm using a 3rd party lib that uses `RCTDeviceEventEmitter`. There are native ways to communicate, but these are not implemented by this iib. With a native subscriber for this events, you would not have to change existing libraries that already using events to communicate with JS.
